From 77b167e36f2a1dc1377952fc958b39dc181425ab Mon Sep 17 00:00:00 2001
From: Henry Chen <henry.chen@oss.cipunited.com>
Date: Tue, 21 May 2024 14:06:17 +0800
Subject: [PATCH] linux-user/mips64/target_elf: detect and use
 EF_MIPS_MACH_LS3A

An ELF file with this flag should be expected to have some
Loongson-3A-exclusive instructions, unsupported by QEMU's
default 5KEf CPU.
---
 linux-user/mips64/target_elf.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/linux-user/mips64/target_elf.h b/linux-user/mips64/target_elf.h
index 5f2f2df29f..35458ff907 100644
--- a/linux-user/mips64/target_elf.h
+++ b/linux-user/mips64/target_elf.h
@@ -15,6 +15,13 @@ static inline const char *cpu_get_model(uint32_t eflags)
     if ((eflags & EF_MIPS_MACH) == EF_MIPS_MACH_5900) {
         return "R5900";
     }
+    if (eflags & EF_MIPS_MACH_LS3A) {
+        if (eflags & EF_MIPS_NAN2008) {
+            return "Loongson-3A4000";
+        } else {
+            return "Loongson-3A1000";
+        }
+    }
     return "5KEf";
 }
 #endif
-- 
2.45.1

